pkg/gateway/gateway_test.go

```diff
@@ -4,11 +4,174 @@ package gateway
+			mocks: func(env *mock_env.MockCore) {
+				env.EXPECT().Environment().AnyTimes().Return(emptyEnv)
+			},
+			expectedAllowList: make([]string, 0),
+			wantErr:           "missing required domain. Ensure the environment has both ActiveDirectoryEndpoint and ResourceManagerEndpoint",
+		},
+		{
+			name:           "GatewayDomains are split, formatted, and added to allowList",
+			gatewayDomains: "doMain1,Domain2",
+			mocks: func(env *mock_env.MockCore) {
+				env.EXPECT().Environment().AnyTimes().Return(&populatedEnv)
+			},
+			expectedAllowList: []string{"domain1", "domain2", "activedirectory", "resourcemanager"},
+		},
+		{
+			name: "Domains in Environments are parsed and appended to allowList",
+			mocks: func(env *mock_env.MockCore) {
+				env.EXPECT().Environment().AnyTimes().AnyTimes().Return(&populatedEnv)
+			},
+			expectedAllowList: []string{"activedirectory", "resourcemanager"},
+		},
+		{
+			name: "Invalid domain format in environment",
+			mocks: func(env *mock_env.MockCore) {
+				invalidEnv := *emptyEnv
+				invalidEnv.ActiveDirectoryEndpoint = "\r"
+				env.EXPECT().Environment().AnyTimes().AnyTimes().Return(&invalidEnv)
+			},
+			wantErr: `parse "\r": net/url: invalid control character in URL`,
+		},
+		{
+			name:          "When the acrResourceID is present it is parsed into domains and added to allowList",
+			acrResourceID: "/subscriptions/93aeba23-2f76-4307-be82-02921df010cf/resourceGroups/global/providers/Microsoft.ContainerRegistry/registries/resourceName",
+			mocks: func(env *mock_env.MockCore) {
+				env.EXPECT().Environment().AnyTimes().Return(&populatedEnv)
+				env.EXPECT().Location().AnyTimes().Return("location")
+			},
+			expectedAllowList: []string{"resourcename.dnssuffix", "resourcename.location.data.dnssuffix", "activedirectory", "resourcemanager"},
+		},
+		{
+			name:          "When the acrResourceID is present but it is invalid",
+			acrResourceID: "invalidAcrResourceID",
+			mocks: func(env *mock_env.MockCore) {
+				env.EXPECT().Environment().AnyTimes().Return(&populatedEnv)
+				env.EXPECT().Location().AnyTimes().Return("location")
+			},
+			wantErr: "parsing failed for invalidAcrResourceID. Invalid resource Id format",
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			controller := gomock.NewController(t)
+			defer controller.Finish()
+
+			baseLog := logrus.NewEntry(logrus.StandardLogger())
+			db := mock_database.NewMockGateway(controller)
+			metrics := mock_metrics.NewMockEmitter(controller)
+
+			httpl, _ := utilnet.Listen("tcp", ":8080", SocketSize)
+			httpsl, _ := utilnet.Listen("tcp", ":8443", SocketSize)
+
+			env := mock_env.NewMockCore(controller)
+			tt.mocks(env)
+
+			gtwy, err := NewGateway(ctx, env, baseLog, baseLog, db, httpsl, httpl, tt.acrResourceID, tt.gatewayDomains, metrics)
+
+			if tt.wantErr != "" {
+				if err == nil {
+					t.Errorf("Expected error: '%s' but did not receive any errors", tt.wantErr)
```

Comment:
nit: `%q` == `'%s'` - this happens in a few places. Feel free to leave as-is.

Reply:
nice! I will update to use this, thank you.